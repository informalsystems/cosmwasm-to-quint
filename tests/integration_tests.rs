use std::{env, fs, path::Path, process::Command, sync::Once};

use anyhow::{ensure, Context, Result};

static SETUP: Once = Once::new();

fn run(dir: &str, f: impl FnOnce(&mut Command)) -> Result<String> {
    let root = env::temp_dir().join("cosmwasm-to-quint");

    let heredir = Path::new(".").canonicalize()?;

    SETUP.call_once(|| {
        let mut cmd = Command::new("cargo");
        cmd.args(["install", "--path", ".", "--debug", "--locked", "--root"]);
        cmd.arg(&root);
        cmd.current_dir(&heredir);
        let status = cmd.status().unwrap();
        if !status.success() {
            panic!("installing cosmwasm-to-quint failed")
        }
    });

    let mut cmd = Command::new("cargo");
    cmd.arg("cosmwasm-to-quint");

    let path = format!(
        "{}:{}",
        root.join("bin").display(),
        env::var("PATH").unwrap_or_else(|_| "".into())
    );
    cmd.env("PATH", path);

    let ws = heredir.join("tests").join("fixtures").join(dir);
    cmd.current_dir(&ws);

    f(&mut cmd);

    let _ = fs::remove_dir_all(ws.join("target"));

    println!("Running {:?}", cmd);
    let output = cmd.output().context("Process failed")?;
    ensure!(
        output.status.success(),
        "Process exited with non-zero exit code. Stderr:\n{}",
        String::from_utf8(output.stderr)?
    );

    Ok(String::from_utf8(output.stdout)?)
}

#[test]
fn ctf01() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-01", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf02() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-02", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf03() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-03", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf04() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-04", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf05() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-05", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf06() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-06", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf07() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-07", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf08() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-08", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf09() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-09", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn ctf10() -> Result<()> {
    let output = run("cosmwasm-ctf/ctf-10", |_cmd| {})?;
    insta::assert_snapshot!(output);
    Ok(())
}
