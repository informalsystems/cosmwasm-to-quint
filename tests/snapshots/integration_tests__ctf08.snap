---
source: tests/integration_tests.rs
expression: output
---
quint:


oaksecurity_cosmwasm_ctf_08_stubs.qnt:
module oaksecurity_cosmwasm_ctf_08 {
  
  import basicSpells.* from "./lib/basicSpells"
  import cw_types.* from "./lib/cw_types"
  import cw_utils.* from "./lib/cw_utils"
  import messaging.* from "./lib/messaging"
  import bank from "./lib/bank"

  
  var contract_state: ContractState
  var result: Result
  var bank: bank::Bank
  var time: int

  
  pure val CONTRACT_ADDRESS = "contract0"

  
  pure val ADDRESSES = Set("sender1", "sender2", "sender3", CONTRACT_ADDRESS)
  pure val DENOMS = Set("d1", "uawesome")
  pure val MAX_AMOUNT = 200


  type InstantiateMsg = { nft_address: str }
  type GetCountResponse = { count: int }
  type Config = { nft_contract: Addr }
  type Sale = { nft_id: str, price: int, owner: Addr, tradable: bool }
  type Trade = { asked_id: str, to_trade_id: str, trader: Addr }
  type Operations = { n_trades: int, n_sales: int }
  type ExecuteMsg =
    | ExecuteMsg_BuyNFT({ id: str })
    | ExecuteMsg_NewSale({ id: str, price: int, tradable: bool })
    | ExecuteMsg_CancelSale({ id: str })
    | ExecuteMsg_NewTrade({ target: str, offered: str })
    | ExecuteMsg_AcceptTrade({ id: str, trader: str })
    | ExecuteMsg_CancelTrade({ id: str })
  pure def instantiate(state: ContractState, _env: Env, _info: MessageInfo, msg: InstantiateMsg): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }

  pure def execute(state: ContractState, env: Env, info: MessageInfo, msg: ExecuteMsg): (Result[Response, ContractError], ContractState) = match msg {
    | ExecuteMsg_BuyNFT(__r) => exec_buy(state, env, info, __r.id)
    | ExecuteMsg_NewSale(__r) => exec_new_sale(state, env, info, __r.id, __r.price, __r.tradable)
    | ExecuteMsg_CancelSale(__r) => exec_cancel_sale(state, info, __r.id)
    | ExecuteMsg_NewTrade(__r) => exec_new_trade(state, env, info, __r.target, __r.offered)
    | ExecuteMsg_AcceptTrade(__r) => exec_accept_trade(state, info, __r.id, __r.trader)
    | ExecuteMsg_CancelTrade(__r) => exec_cancel_trade(state, info, __r.id)
  }

  pure def exec_new_sale(state: ContractState, env: Env, info: MessageInfo, id: str, price: int, tradable: bool): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_new_sale_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_id: str = Set("s1", "s2", "s3").oneOf()
    nondet message_price: int = 0.to(MAX_AMOUNT).oneOf()
    nondet message_tradable: bool = Bool.oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_NewSale({ id: message_id, price: message_price, tradable: message_tradable })
    execute_message(message, max_funds)
  }

  pure def exec_buy(state: ContractState, _env: Env, info: MessageInfo, id: str): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_buy_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_id: str = Set("s1", "s2", "s3").oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_BuyNFT({ id: message_id })
    execute_message(message, max_funds)
  }

  pure def exec_cancel_sale(state: ContractState, info: MessageInfo, id: str): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_cancel_sale_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_id: str = Set("s1", "s2", "s3").oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_CancelSale({ id: message_id })
    execute_message(message, max_funds)
  }

  pure def exec_new_trade(state: ContractState, env: Env, info: MessageInfo, asked_id: str, offered_id: str): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_new_trade_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_target: str = Set("s1", "s2", "s3").oneOf()
    nondet message_offered: str = Set("s1", "s2", "s3").oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_NewTrade({ target: message_target, offered: message_offered })
    execute_message(message, max_funds)
  }

  pure def exec_accept_trade(state: ContractState, info: MessageInfo, asked_id: str, trader: str): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_accept_trade_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_id: str = Set("s1", "s2", "s3").oneOf()
    nondet message_trader: str = Set("s1", "s2", "s3").oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_AcceptTrade({ id: message_id, trader: message_trader })
    execute_message(message, max_funds)
  }

  pure def exec_cancel_trade(state: ContractState, info: MessageInfo, asked_id: str): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }
                            
  action exec_cancel_trade_action = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = MAX_AMOUNT
    nondet message_id: str = Set("s1", "s2", "s3").oneOf()
    pure val message: ExecuteMsg = ExecuteMsg_CancelTrade({ id: message_id })
    execute_message(message, max_funds)
  }
  pure def reply(state: ContractState, _env: Env, reply: Reply): (Result[Response, ContractError], ContractState) = {
    // TODO: Update body
    (Ok(Response_new), state)
  }

  pure val DENOM = "uawesome"
  pure val TRADE_REPLY = 1
  pure val SALE_REPLY = 2

  type ContractState = {
    config: Config,
    sales: str -> Sale,
    trades: str -> Trade,
    operations: Operations
  }

  pure val init_contract_state = {
    config: { nft_contract: "s1" },
    sales: Map(),
    trades: Map(),
    operations: { n_trades: 0,n_sales: 0 }
  }

  action execute_step = all {
    any {
      exec_new_sale_action,
      exec_buy_action,
      exec_cancel_sale_action,
      exec_new_trade_action,
      exec_accept_trade_action,
      exec_cancel_trade_action
    },
    advance_time,
  }



  pure val init_bank_state = ADDRESSES.mapBy(_ => DENOMS.mapBy(_ => MAX_AMOUNT))

  val env_val = { block: { time: time } }

  action init = {
    // TODO: Change next line according to fund expectations
    pure val max_funds = 0

    nondet sender = Set("admin").oneOf()
    nondet denom = DENOMS.oneOf()
    nondet amount = 0.to(max_funds).oneOf()
    val funds = [{ denom: denom, amount: amount }]
    val info = { sender: sender, funds: funds }

    pure val message: InstantiateMsg = { nft_address: "" }
    pure val r = instantiate(init_contract_state, { block: { time: 0 } }, info, message)

    all {
      contract_state' = r._2,
      bank' = init_bank_state,
      result' = r._1,
      time' = 0,
    }
  }


  action execute_message(message, max_funds) = {
    nondet sender = ADDRESSES.oneOf()
    nondet denom = DENOMS.oneOf()
    nondet amount = 0.to(max_funds).oneOf()
    val funds = [{ denom: denom, amount: amount }]
    val info = { sender: sender, funds: funds }

    val r = execute(contract_state, env_val, info, message)
    all {
      bank.get(sender).get(denom) >= amount,
      match r._1 {
        | Ok(_) => bank' = bank.setBy(sender, balances => balances.setBy(denom, balance => balance - amount))
                               .setBy(CONTRACT_ADDRESS, balances => balances.setBy(denom, balance => balance + amount))
        | Err(_) => bank' = bank
      },
      result' = r._1,
      contract_state' = r._2,
    }
  }

  action advance_time = time' = time + 1

  action step = {
    val message_getting = get_message(result)
    val new_result = message_getting._1
    val opt_message = message_getting._2
    match opt_message {
      | Some(submsg) => {
          val current_state = { bank: bank, result: new_result, contract_state: contract_state }
          val new_state = process_message(current_state, env_val, CONTRACT_ADDRESS, submsg, reply)
          all {
            bank' = new_state.bank,
            result' = new_state.result,
            contract_state' = new_state.contract_state,
            advance_time,
          }
      }
      | None => execute_step
    }
  }

}


mbt:
mbt_oaksecurity_cosmwasm_ctf_08.rs:

pub mod state_structs {
    use num_bigint::BigInt;
    use serde::Deserialize;
    use std::collections::HashMap;
    use itf::de::{self, As};
    
    #[derive(Clone, Debug, Deserialize)]
    pub struct Config {
        pub nft_contract: String
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct GetCountResponse {
        pub count: BigInt
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct InstantiateMsg {
        pub nft_address: String
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct Operations {
        pub n_trades: BigInt,
        pub n_sales: BigInt
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct Sale {
        pub nft_id: String,
        pub price: BigInt,
        pub owner: String,
        pub tradable: bool
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct Trade {
        pub asked_id: String,
        pub to_trade_id: String,
        pub trader: String
    }
    
    #[derive(Clone, Debug, Deserialize)]
    pub struct ContractState {
        pub config: Config,
        pub sales: HashMap<String, Sale>,
        pub trades: HashMap<String, Trade>,
        pub operations: Operations
    }
    
    #[derive(Clone, Debug, Deserialize)]
    pub struct NondetPicks {
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub sender: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub denom: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub amount: Option<BigInt>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_id: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_price: Option<BigInt>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_tradable: Option<bool>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_id: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_id: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_target: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_offered: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_id: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_trader: Option<String>,
        
        #[serde(with = "As::<de::Option::<_>>")]
        pub message_id: Option<String>
    }
    
    #[derive(Clone, Debug, Deserialize)]
    pub struct Message {}
    #[derive(Clone, Debug, Deserialize)]
    pub struct Attribute {
        pub key: String,
        pub value: QuintSerializedValue,
    }

    #[derive(Clone, Debug, Deserialize)]
    #[serde(tag = "tag", content = "value")]
    pub enum QuintSerializedValue {
        FromInt(BigInt),
        FromStr(String),
        FromListInt(Vec<BigInt>),
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct Response {
        pub messages: Vec<Message>,
        pub attributes: Vec<Attribute>,
    }

    #[derive(Clone, Debug, Deserialize)]
    pub struct State {
        pub contract_state: ContractState,
        pub bank: HashMap<String, HashMap<String, BigInt>>,
        #[serde(with = "As::<de::Result::<_, _>>")]
        pub result: Result<Response, String>,
        pub action_taken: String,
        pub nondet_picks: NondetPicks,
        pub time: BigInt,
    }

}
    
#[cfg(test)]
pub mod tests {
    use oaksecurity_cosmwasm_ctf_08::contract;
    use oaksecurity_cosmwasm_ctf_08::msg::{ExecuteMsg, InstantiateMsg};


    use crate::state_structs::*;
    use cosmwasm_std::{coin, Addr, Uint128};
    use cw_multi_test::{App, AppResponse, ContractWrapper, Executor};
    use itf::trace_from_str;
    use num_bigint::BigInt;
    use num_traits::{ToPrimitive, Zero};

    pub const DENOM: &str = "uawesome";
    pub const TICK: u64 = 1;

    pub fn mint_tokens(mut app: App, recipient: String, denom: String, amount: Uint128) -> App {
        app.sudo(cw_multi_test::SudoMsg::Bank(
            cw_multi_test::BankSudo::Mint {
                to_address: recipient.to_owned(),
                amount: vec![coin(amount.u128(), denom)],
            },
        ))
        .unwrap();
        app
    }

    fn compare_state(test_state: &TestState, app: &App, state: &State) {
        // compare contract balances
        let balance = app
            .wrap()
            .query_balance(&test_state.contract_addr, DENOM)
            .unwrap()
            .amount;
        let trace_balance = state
            .bank
            .get(&test_state.contract_addr.to_string())
            .and_then(|x| x.get(DENOM))
            .and_then(|x| x.to_u128())
            .unwrap_or(0);
        println!(
            "Contract balance ({:?}) for {DENOM}: {:?} vs {:?}",
            test_state.contract_addr,
            balance,
            Uint128::new(trace_balance)
        );
        assert_eq!(balance, Uint128::new(trace_balance));

        // TODO: Query the contract and compare the state as you wish
    }

    fn compare_result(
        trace_result: Result<Response, String>,
        app_result: Result<AppResponse, anyhow::Error>,
    ) {
        if trace_result.is_ok() {
            assert!(
                app_result.is_ok(),
                "Action unexpectedly failed, error: {:?}",
                app_result.err()
            );
            println!("Action successful as expected");
        } else {
            assert!(
                app_result.is_err(),
                "Expected action to fail with error: {:?}",
                trace_result.err()
            );
            println!("Action failed as expected");
        }
    }

    fn funds_from_trace(amount: Option<BigInt>, denom: Option<String>) -> Vec<cosmwasm_std::Coin> {
        if amount.is_none() || denom.is_none() || amount == Some(Zero::zero()) {
            return vec![];
        }

        vec![coin(
            amount.as_ref().unwrap().to_u128().unwrap(),
            denom.unwrap(),
        )]
    }

    // Testing is stateful.
    struct TestState {
        // we will only know the contract address once we have processed an `instantiate` step
        pub contract_addr: Addr,
    }

    #[test]
    fn model_test() {
        let mut app = App::default();
        let code = ContractWrapper::new(contract::execute, contract::instantiate, contract::query);
        let code_id = app.store_code(Box::new(code));

        // create test state
        let mut test_state = TestState {
            contract_addr: Addr::unchecked("contract0"),
        };

        // load trace data
        let data = include_str!("../quint/test.itf.json");
        let trace: itf::Trace<State> = trace_from_str(data).unwrap();

        for s in trace.states {
            let last_result = s.value.result.clone();
            if last_result.is_ok() && !last_result.unwrap().messages.is_empty() {
                println!("Processing messages, skipping");
                continue;
            }

            let action_taken = &s.value.action_taken;
            let nondet_picks = &s.value.nondet_picks;
            let amount = nondet_picks.amount.clone();
            let denom = nondet_picks.denom.clone();
            let sender = nondet_picks.sender.clone();

            println!("Step number: {:?}", s.meta.index);
            println!("Result from trace: {:?}", s.value.result.clone());

            match action_taken.as_str() {


               "exec_accept_trade_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_id = nondet_picks.message_id.clone().unwrap();
                    let message_trader = nondet_picks.message_trader.clone().unwrap();
                    let msg = ExecuteMsg::AcceptTrade { id: message_id, trader: message_trader };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "exec_buy_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_id = nondet_picks.message_id.clone().unwrap();
                    let msg = ExecuteMsg::BuyNFT { id: message_id };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "exec_cancel_sale_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_id = nondet_picks.message_id.clone().unwrap();
                    let msg = ExecuteMsg::CancelSale { id: message_id };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "exec_cancel_trade_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_id = nondet_picks.message_id.clone().unwrap();
                    let msg = ExecuteMsg::CancelTrade { id: message_id };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "exec_new_sale_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_id = nondet_picks.message_id.clone().unwrap();
                    let message_price = nondet_picks.message_price.clone().unwrap().to_u64().unwrap().into();
                    let message_tradable = nondet_picks.message_tradable.clone().unwrap();
                    let msg = ExecuteMsg::NewSale { id: message_id, price: message_price, tradable: message_tradable };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "exec_new_trade_action" => {
                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let message_target = nondet_picks.message_target.clone().unwrap();
                    let message_offered = nondet_picks.message_offered.clone().unwrap();
                    let msg = ExecuteMsg::NewTrade { target: message_target, offered: message_offered };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    let res = app.execute_contract(
                        sender,
                        test_state.contract_addr.clone(),
                        &msg,
                        &funds,
                    );

                    compare_result(s.value.result.clone(), res)
                }


               "q::init" => {
                    println!("Initializing contract.");

                    let sender = Addr::unchecked(sender.unwrap());
                    let funds = funds_from_trace(amount, denom);

                    let msg = InstantiateMsg { nft_address: "" };
                    println!("Message: {:?}", msg);
                    println!("Sender: {:?}", sender);
                    println!("Funds: {:?}", funds);

                    test_state.contract_addr = app.instantiate_contract(
                        code_id,
                        sender,
                        &msg,
                        &funds,
                        "test",
                        None,
                    ).unwrap();

                    for (addr, coins) in s.value.bank.clone().iter() {
                        for (denom, amount) in coins.iter() {
                            app = mint_tokens(
                                app,
                                addr.clone(),
                                denom.to_string(),
                                Uint128::new(amount.to_u128().unwrap()),
                            );
                        }
                    }

               }

                _ => panic!("Invalid action taken"),
            }
            compare_state(&test_state, &app, &(s.value.clone()));
            println!(
                "clock is advancing for {} seconds",
                TICK
            );
            app.update_block(|block| {
                block.time = block.time.plus_seconds(TICK);
            });
            println!("-----------------------------------");
        }
    }
}
